FSDP + Tensor Parallelism

erhoo82 opened the pull request:
# What does this PR do ?

This supports torch FSDP training with Tensor-parallelism.

Set `model.fsdp=True` to enable FSDP. Gradient reduction data type can be set as `model.fsdp_grad_reduce_dtype=bf16`.

FSDP mode is not compatible with the Distributed Optimizer and O2 optimizer wrapper, thus should set `model.megatron_amp_O2=False` and not use `model.optim.name=distributed_fused_adam`.

To store checkpoint in FSDP sharded format, set model.fsdp_sharded_checkpoint=True.

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.


## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

athitten commented:
jenkins

athitten commented:
jenkins

ericharper commented:
jenkins

stu1130 commented:
@erhoo82 Thanks for hard working. Do you have timeline in mind when the PR will get merged? Thanks!

erhoo82 commented on nemo/collections/nlp/parts/nlp_overrides.py:
```diff
@@ -448,6 +468,329 @@ def _configure_launcher(self):
+
+        super().__init__(**kwargs)
+
+    def _set_mixed_precision_recipe(
```
When are we switching to PTL 2.0.7?
Yes, need to be in sync with the latest PTL when switching the version.

athitten replied:
We are already using PTL 2.0.7 in NeMo.

athitten replied:
@erhoo82 shouldn't this be fixed ? `_set_mixed_precision_recipe` doesn't exist in PTL 2.0.7.

erhoo82 replied:
I think this should be fine. This is still a standalone function I define here. Even if the later PTL redefines it, it should be overridden by this function in NeMo code.

erhoo82 commented:
@stu1130 Have been caught up by priority work. Trying to close soon.

erhoo82 commented:
@athitten,  @ericharper 
- I have removed the torch FSDP patch file as the bug was fixed at PyT23.09 container.
- I have fixed some confusion in the mixed precision definition. Currently, NeMo sets the master param data type as FP32 when `megatron_O2=False`, which is the case of FSDP. I removed explicit model data type casting to FP32 after FSDP because it is already FP32 when `megatron_O2=False`. Also, `param_dtype` in FSDP sets only the compute data type in forward and backward propagation, which matches the original intention of FSDP design.

athitten commented:
jenkins

erhoo82 commented:
`ModuleNotFoundError: No module named 'transformer_engine'`

athitten commented:
jenkins

athitten commented:
jenkins

athitten commented on nemo/collections/nlp/parts/nlp_overrides.py:
```diff
@@ -64,17 +82,55 @@
         optim_state_to_sharding_state,
     )
     from megatron.core.transformer.module import Float16Module as MCoreFloat16Module
+    from megatron.core.transformer.transformer_layer import TransformerLayer as MCoreTransformerLayer
```
@erhoo82 @ericharper do we want to move this to a separate try except block ? Cause we would have `HAVE_MEGATRON_CORE=False` if the import fails with `TransformerEngine not found error`, although megatron core would be available and can cause error for blocks that depend just on mcore and not TE.

ericharper replied:
We can leave it as is for now. There is a separate  effort to improve the import guarding.

athitten commented:
jenkins

ericharper commented on nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:
```diff
@@ -728,6 +731,21 @@ def allreduce_sequence_parallel_gradients(self):
         for buf, synced in zip(grads, torch._utils._unflatten_dense_tensors(coalesced, grads)):
             buf.copy_(synced)
 
+    def allreduce_fsdp_sharding_omitted_gradients(self):
```
Doesn't have to be this PR, but shouldn't these calls go into an fsdp strategy rather than the model code? Maybe @athitten , you can address this is in a future PR.

ericharper commented on nemo/collections/nlp/modules/common/megatron/clip_grads.py:
```diff
@@ -54,7 +54,8 @@
 
-def clip_grad_norm_fp32(parameters, max_norm, norm_type=2):
+@torch.no_grad()
+def clip_grad_norm_fp32(parameters, max_norm, norm_type=2, use_fsdp=False):
```
@athitten , we should be getting this from mcore. (for future pr)

ericharper commented on nemo/collections/nlp/parts/nlp_overrides.py:
```diff
@@ -451,6 +468,299 @@ def _configure_launcher(self):
+    """ FSDP plugin for Pytorch Lightning with the support for tensor-parallelism.
+
+    Args:
+        sharding_strategy: FSDP parameter sharding strategy.
+        grad_reduce_dtype: Data type for FSDP gradient shard ReduceScatter.
+        sharded_checkpoint: Store/load FSDP-sharded checkpoints.
+        precision: Precision recipe to be used with FSDP.
```
Can go in a future PR, but I'd like to see more detail here. Like what are the possible values and also links to more information.

ericharper commented on nemo/collections/nlp/parts/nlp_overrides.py:
```diff
@@ -451,6 +468,299 @@ def _configure_launcher(self):
+        shared by FSDP.
+        `reduce_dtype` sets gradient reduction data type.
+        """
+        if precision in ["16-true", "16-mixed", 16]:
+            param_dtype = reduce_dtype = buffer_dtype = torch.float16
+        elif precision in ["bf16-true", "bf16-mixed", "bf16"]:
+            param_dtype = reduce_dtype = buffer_dtype = torch.bfloat16
+        elif precision == 32:
+            param_dtype = reduce_dtype = buffer_dtype = torch.float
+        else:
+            raise ValueError(f"Was unable to infer precision type, received {precision!r}.")
```
Again for a future PR but we need to use a utility function for this that can be shared across all of nemo.

ericharper commented on nemo/collections/nlp/parts/nlp_overrides.py:
```diff
@@ -451,6 +468,299 @@ def _configure_launcher(self):
+    return state_dict_type_context
+
+
+class NLPFSDPStrategy(FSDPStrategy):
```
Also noting here for a future PR that nlp_overrides.py is too long now and which refactor it a bit.

ericharper reviewed (approved):
LGTM. Thanks!
